File: 24-1-计算机网络-潘俊杉/窗口滑动.py
from random import randint, choice, triangular

class TransportLayer:
    def __init__(self, window_size):
        self.send_window = window_size                  # 发送窗口大小
        self.receive_window = window_size               # 接收窗口大小
        self.front = window_size + 1                    # 窗口前沿（右边）
        self.mid = 0                                    # 中间指示已发送但未确认的位置
        self.back = 0                                   # 窗口后沿（左边）
        self.send_buffer = [None] * window_size         # 发送缓冲区
        self.receive_buffer = [None] * window_size      # 接收缓冲区
        self.received = [False] * window_size           # 标记接收缓冲区是否接收到数据

    def send(self, datas):
        if self.mid == self.front:
            print("窗口已满，等待 ACK")
            return
        for data in datas:
            self.send_buffer[self.mid - self.back] = data
            print(f"发送数据：{data} 序列号：{self.mid}")
            self.mid += 1

    # 返回 ack
    def receive(self, datas):
        flag = False
        for i in range(len(datas)):
            if not self.receive_buffer[i]:
                select = choice([True, False])
                if select:
                    flag = True
                    self.receive_buffer[i] = datas[i]
                    self.received[i] = True
                    print(f"接收数据：{datas[i]}")

        if not flag:
            print("没有接收到任何数据!")
            return self.back
        for i in range(self.receive_window):
            if not self.received[i]:
                return self.back + i
        return self.front
    
    def submit(self, ack):
        if self.back == ack:
            print(f"没有可以提交的数据!")
        for i in range(self.back, ack):
            print(f"向上提交数据：{self.receive_buffer[0]}")
            self.receive_buffer.pop(0)
            self.received.pop(0)
            self.receive_buffer.append(None)
            self.received.append(False)
        
        rwnd = round(triangular(self.receive_window - 2, self.receive_window + 2, self.receive_window))  # 三角分布(下界，上界，峰值)
        self.receive_buffer = self.receive_buffer[:rwnd] if rwnd < self.receive_window else self.receive_buffer + [None] * (rwnd - self.receive_window)
        self.received = self.received[:rwnd] if rwnd < self.receive_window else self.received + [False] * (rwnd - self.receive_window)
        self.receive_window = rwnd
        print(f"接收窗口大小变为：{self.receive_window}")
        return rwnd

    def slide_window(self, ack, rwnd):
        for i in range(self.back, ack):
            self.send_buffer.pop(0)
            self.send_buffer.append(None)
            self.back += 1
            self.front += 1
        
        self.mid = self.back
        self.send_buffer = self.send_buffer[:rwnd] if rwnd < self.send_window else self.send_buffer + [None] * (rwnd - self.send_window)
        self.send_window = rwnd
        print(f"发送窗口大小变为：{self.send_window}")


R = 10              # 交互次数
window_size = 10    # 初始窗口大小
transport_layer = TransportLayer(window_size)

for i in range(R):
    print(f"第{i + 1}次交互".center(35, '-'))
    datas = list(range(transport_layer.back, transport_layer.back + randint(1, transport_layer.send_window)))
    transport_layer.send(datas)
    ack = transport_layer.receive(datas)
    rwnd = transport_layer.submit(ack)
    transport_layer.slide_window(ack, rwnd)
    print(transport_layer.send_buffer)
    print(transport_layer.receive_buffer)

File: 23-24-2-计算机网络-潘俊杉/窗口滑动.cpp
#include <iostream>
#include <vector>
#include <random>

class TransportLayer
{
public:
    int send_window;                 // 发送窗口大小
    int receive_window;              // 接收窗口大小
    int front;                       // 窗口前沿（右边）
    int mid;                         // 中间指示已发送但未确认的位置
    int back;                        // 窗口后沿（左边）
    std::vector<int> send_buffer;    // 发送缓冲区
    std::vector<int> receive_buffer; // 接收缓冲区
    std::vector<bool> received;      // 标记接收缓冲区是否接收到数据


    TransportLayer(int window_size)
    {
        send_window = window_size;
        receive_window = window_size;
        front = window_size + 1;
        mid = 0;
        back = 0;
        send_buffer.resize(window_size);
        receive_buffer.resize(window_size);
        received.resize(window_size, false);
    }

    void send(std::vector<int> datas)
    {
        if (mid == front)
        {
            std::cout << "窗口已满，等待 ACK" << std::endl;
            return;
        }
        for (int data : datas)
        {
            send_buffer[mid - back] = data;
            std::cout << "发送数据：" << data << " 序列号：" << mid << std::endl;
            mid++;
        }
    }

    int receive(std::vector<int> datas)
    {
        bool flag = false;
        for (int i = 0; i < datas.size(); ++i)
        {
            if (!receive_buffer[i])
            {
                bool select = rand() % 2 == 0; // choice([True, False])
                if (select)
                {
                    flag = true;
                    receive_buffer[i] = datas[i];
                    received[i] = true;
                    std::cout << "接收数据：" << datas[i] << std::endl;
                }
            }
        }
        if (!flag)
        {
            std::cout << "没有接收到任何数据!" << std::endl;
            return back;
        }
        for (int i = 0; i < receive_window; ++i)
        {
            if (!received[i])
            {
                return back + i;
            }
        }
        return front;
    }

    int submit(int ack)
    {
        if (back == ack)
        {
            std::cout << "没有可以提交的数据!" << std::endl;
        }
        for (int i = back; i < ack; ++i)
        {
            std::cout << "向上提交数据：" << receive_buffer[0] << std::endl;
            receive_buffer.erase(receive_buffer.begin());
            received.erase(received.begin());
            receive_buffer.push_back(0);
            received.push_back(false);
        }
        int rwnd = round((receive_window - 2 + receive_window + 2 + receive_window) / 3.0);
        receive_buffer.resize(rwnd < receive_window ? rwnd : receive_window, 0);
        received.resize(rwnd < receive_window ? rwnd : receive_window, false);
        receive_window = rwnd;
        std::cout << "接收窗口大小变为：" << receive_window << std::endl;
        return rwnd;
    }

    void slide_window(int ack, int rwnd)
    {
        for (int i = back; i < ack; ++i)
        {
            send_buffer.erase(send_buffer.begin());
            send_buffer.push_back(0);
            back++;
            front++;
        }
        mid = back;
        send_buffer.resize(rwnd < send_window ? rwnd : send_window, 0);
        send_window = rwnd;
        std::cout << "发送窗口大小变为：" << send_window << std::endl;
    }
};

int main()
{
    const int R = 10;           // 交互次数
    const int window_size = 10; // 初始窗口大小
    TransportLayer transport_layer(window_size);

    for (int i = 0; i < R; ++i)
    {
        std::cout << std::string(35, '-') << std::endl;
        std::cout << "第" << i + 1 << "次交互" << std::endl;
        std::vector<int> datas;
        int start = transport_layer.back;
        int end = transport_layer.back + (rand() % transport_layer.send_window) + 1;
        for (int j = start; j < end; ++j)
            datas.push_back(j);
        transport_layer.send(datas);
        int ack = transport_layer.receive(datas);
        int rwnd = transport_layer.submit(ack);
        transport_layer.slide_window(ack, rwnd);
        for (int num : transport_layer.send_buffer)
            std::cout << num << " ";
        std::cout << std::endl;
        for (int num : transport_layer.receive_buffer)
            std::cout << num << " ";
        std::cout << std::endl;
    }

    return 0;
}
